### PR TITLE
Various updates to subsys/mgmt/osdp

### PR DESCRIPTION
**Pull request by sidcha**
This PR brings in various updates such as bug fixes, cleanups and new features to subsys/mgmt/osdp in both control panel and peripheral device modes. These changes have been tested with a st-nucleo-g474re board with secure channel enabled.

Although this PR does way more than any single PR should be doing, I've taken care to split these changes across 30 commits to make them more digestible.

### CODE REVIEW

File: MAINTAINERS.yml
Comment (stephanosio) [inline MAINTAINERS.yml:1225]: > subsys/mgmt/osdp _has_ a dedicated maintainer who will rush to address
any issues identified :).

In that case, please add yourself as a maintainer.

Comment (sidcha) [reply MAINTAINERS.yml]: Done.

File: include/zephyr/mgmt/osdp.h
Comment (adakus) [inline include/zephyr/mgmt/osdp.h:462]: I guess this one should be outside of #ifdef CONFIG_OSDP_SC_ENABLED ... #endIf

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (adakus) [inline subsys/mgmt/osdp/src/osdp_pd.c:981]: For CONFIG_OSDP_SC_ENABLED == 0 struct member  'sc_tstamp' is not visible

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (adakus) [inline subsys/mgmt/osdp/src/osdp_pd.c:672]: According to spec this status can take values in range: 0-2

File: include/zephyr/mgmt/osdp.h
Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:462]: Thanks for the review, I disabled CONFIG_OSDP_SC_ENABLED and fixed this and other issues that were caused due to code leaks our side this switch.

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:981]: Fixed (see above).

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:672]: Right. But it is the current limitation of the implementation and this patch-set is not doing anything about it. We can take this up as a separate task once this PR is merged.

File: include/zephyr/mgmt/osdp.h
Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:462]: Also, CONFIG_OSDP_SC_ENABLED situation has worsened now more than ever. So I'll add another patch to reduce the number of  ifdefs we use to disable secure channel.

**Comment by sidcha** [comment]
I think the 4 checkpatch issues due to the following are valid exceptions to that rule.

```
-:1339: WARNING:MACRO_WITH_FLOW_CONTROL: Macros with flow control statements should be avoided
#1339: FILE: subsys/mgmt/osdp/src/osdp_cp.c:133:
+	#define ASSERT_BUF_LEN(need)                                         \
+		if (max_len < need) {                                        \
+			LOG_ERR("OOM at build CMD(%02x) - have:%d, need:%d", \
+				pd->cmd_id, max_len, need);                  \
+			return OSDP_CP_ERR_GENERIC;                          \
+		}
```

**Comment by gmarull** [comment]
> I think the 4 checkpatch issues due to the following are valid exceptions to that rule.
> 
> ```
> -:1339: WARNING:MACRO_WITH_FLOW_CONTROL: Macros with flow control statements should be avoided
> #1339: FILE: subsys/mgmt/osdp/src/osdp_cp.c:133:
> +	#define ASSERT_BUF_LEN(need)                                         \
> +		if (max_len < need) {                                        \
> +			LOG_ERR("OOM at build CMD(%02x) - have:%d, need:%d", \
> +				pd->cmd_id, max_len, need);                  \
> +			return OSDP_CP_ERR_GENERIC;                          \
> +		}
> ```

I agree with checkpatch here, a macro named "ASSERT" should just assert, not obfuscate logic.

**Comment by sidcha** [comment]
> I agree with checkpatch here, a macro named "ASSERT" should just assert, not obfuscate logic

The name is misleading then, I can change it to something more appropriate. These conditions can happen at run time due to elevated activity or due to a malformed command/response; if so we send out an error response.

**Comment by sidcha** [comment]
@gmarull, I've removed the use of the macro by converting half of them into plan asserts (where we control things) and the other half to literal code substitutions due to lack of other options.

Now all tests are passing!

**Comment by carlescufi** [comment]
@adakus please take another look

**Review by adakus — DISMISSED** [review]

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:302]: Oops! Some rebase artefact. Fixed, will be fixed in next version.

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (sidcha) [reply subsys/mgmt/osdp/src/osdp_pd.c:184]: Added checks on cmd where it is used. Thanks.

**Comment by adakus** [comment]
Hi,
I have bug in your implementation

***@***.***

When SC is not enabled then:

***@***.***

Only one item is copied – for second item function code is equal to 0.
So must be fixed – otherwise other capabilities are not set using config defines
Best regards,
Adam

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

From: Siddharth ***@***.***>
Sent: Thursday, July 7, 2022 10:40 PM
To: ***@***.***>
Cc: Kusiek, ***@***.***>; ***@***.***>
Subject: [EXT] Re: [zephyrproject-rtos/zephyr] Various updates to subsys/mgmt/osdp (PR #46449)

*** Please use caution this is an externally originating email. ***

@sidcha commented on this pull request.


In subsys/mgmt/osdp/src/osdp_pd.c<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzephyrproject-rtos%2Fzephyr%2Fpull%2F46449%23discussion_r916263444&data=05%7C01%7Cadam.kusiek1%40hidglobal.com%7C634b266e000746cb79b908da6058dc1c%7Cf0bdc1c951484f86ac40edd976e1814c%7C0%7C0%7C637928232101044637%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=4ZNAF1Ea8Xp4cdYZ0fsUIO6e4ksuKZqaw0cri%2BrTNCI%3D&reserved=0>:

> +     case OSDP_EVENT_KEYPRESS:

+              reply_code = REPLY_KEYPPAD;

+              break;

+       default:

+              LOG_ERR("Unknown event type %d", event->type);

+              break;

+       }

+       if (reply_code == 0) {

+              /* POLL command cannot fail even when there are errors here */

+              return REPLY_ACK;

+       }

+       memcpy(pd->ephemeral_data, event, sizeof(struct osdp_event));

+       return reply_code;

+}

+

+static int pd_cmd_cap_ok(struct osdp_pd *pd, struct osdp_cmd *cmd)

Added checks on cmd where it is used. Thanks.

File: subsys/mgmt/osdp/src/osdp_pd.c
Comment (sidcha) [inline subsys/mgmt/osdp/src/osdp_pd.c:72]: This hunk should not be here, will be removed in next revision. Thanks @adakus.

**Comment by sidcha** [comment]
Can we have some eyes on this one please :)

cc: @gmarull, @carlescufi, @adakus, @stephanosio.

**Comment by sidcha** [comment]
@galak, can you look at this PR as well :)

**Comment by r2r0** [comment]
@MaureenHelm I ask you very kindly as a code owner - could you please consider reviewing this PR or discuss topic who is in charge to do that?
It looks that it is somewhat catch 22 - no owner then no review then no owner assignement then no review ...
We (as HID company) use this code in production environment and it will help us a lot to have it in main branch.

**Review by stephanosio — DISMISSED** [review]
I haven't done an in-depth review of all the changes, but they look reasonable at first glance.

File: include/zephyr/mgmt/osdp.h
Comment (adakus) [inline include/zephyr/mgmt/osdp.h:296]: Relevant events/cmds should have the same size of members. 
This comes out from the fact that event can be used as response on POLL cmd. 
In this situation event (provided in ephemeral data) is directly casted to relevant command struct so memory layout of cmd/event structs should be the same.

Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:296]: Thanks for the review. In PD, `CMD_MFG` copies the response from the app as `struct osdp_cmd` into ephemeral_data ([here](https://github.com/goToMain/zephyr/blob/osdp/subsys/mgmt/osdp/src/osdp_pd.c#L499)) and in `REPLY_MFGREP`, we cast it back to `struct osdp_cmd`.

So there is no problem with that code as it is now but still, I agree it can be a problem in future if we want to allow PDs to report unsolicited MFGREPs. I'll make `osdp_event_mfgrep` look like `osdp_cmd_mfg`

**Comment by MaureenHelm** [comment]
> Although this PR does way more than any single PR should be doing, I've taken care to split these changes across 30 commits to make them more digestible.

It's a lot of changes to review in one PR. You'll have better success getting reviews if you submit smaller batches of commits in separate PRs.

I've also added the `dev-review` label to put this PR on the meeting agenda for this week and find other stakeholders that can help review your changes. You're invited to join the meeting:
https://github.com/zephyrproject-rtos/zephyr/wiki/Zephyr-Committee-and-Working-Group-Meetings#zephyr-dev-meeting

**Comment by sidcha** [comment]
> > Although this PR does way more than any single PR should be doing, I've taken care to split these changes across 30 commits to make them more digestible.
> 
> It's a lot of changes to review in one PR. You'll have better success getting reviews if you submit smaller batches of commits in separate PRs.

I know, I'm sorry. But I can explain :) 

> I've also added the `dev-review` label to put this PR on the meeting agenda for this week and find other stakeholders that can help review your changes. You're invited to join the meeting: https://github.com/zephyrproject-rtos/zephyr/wiki/Zephyr-Committee-and-Working-Group-Meetings#zephyr-dev-meeting

Thank you, I'll join the call.

**Comment by sidcha** [comment]
@adakus, change you suggested is incorporated. Can you take a look again and approve this PR please?

@stephanosio, after your last review only a small fixup (above) was applied. Please take another look.

**Comment by sidcha** [comment]
@r2r0, please take another look.

**Review by adakus — DISMISSED** [review]

**Review by stephanosio — DISMISSED** [review]

**Comment by sidcha** [comment]
Thank you @adakus  and @stephanosio

**Comment by sidcha** [comment]
@MaureenHelm we have 2 reviews now :)

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thank you for organizing and documenting your changes in a nice series of small atomic commits. However, public API and incorrect copyright changes are buried under a lot of implementation cleanup changes. I don't see anything controversial except for the copyright, but I'd like the public API changes submitted in a separate PR to give them a dedicated review.

File: include/zephyr/mgmt/osdp.h
Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:2]: Having them both would defat the point (adding more confusion). I'll drop this change in the next revision.

File: include/zephyr/mgmt/osdp.h
Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:229]: There are currently 8 commits that can touch osdp.h:
```
23e778e38a mgmt/osdp: Reduce ifdefs needed to disable secure channel
c0177e7a58 mgmt/osdp: Change email address to gmail
2b38d1f370 mgmt/osdp: Add support for manufacturer command/reply
0d8ef2e529 mgmt/osdp: Fix doxygen comment formatting in osdp.h
03f63a30c8 mgmt/osdp: Add API to get status of all PDs
79ff7fd9f7 mgmt/osdp: Extend get_sc_status_mask to handle more than 32 PDs
17fa794d13 mgmt/osdp: api: Add support for command complete callback
1fe9943081 mgmt/osdp: Add support for event delivery and notifications
```
I don't know if they can be cleanly separated (some patches lead into the header medication patch). If this cannot be done, I'll just split the entire series into smaller PRs as suggested in dev-meeting and close this one.

File: include/zephyr/mgmt/osdp.h
Comment (stephanosio) [reply include/zephyr/mgmt/osdp.h:2]: IMHO, this is not really a problem. This is not a "copyright" change; it is just updating an outdated email address.

Comment (sidcha) [reply include/zephyr/mgmt/osdp.h:2]: @MaureenHelm can should I drop or keep this change?

**Comment by sidcha** [comment]
@MaureenHelm I don't think splitting this PR is of much help; much let me explain. I just created #49302 with one logical block "introducing events: (it has an API change).  The rest of the 4 blocks ([2/5](https://github.com/goToMain/zephyr/tree/osdp_pd_cleanups), [3/5](https://github.com/goToMain/zephyr/tree/osdp_cp_cleanups), [4/5](https://github.com/goToMain/zephyr/tree/osdp_api_cleanups), [5/5](https://github.com/goToMain/zephyr/tree/osdp_misc_fixes)) depend on their predecessors' changes so they cannot be based on the main branch. If I create all 5 PRs at once, 5/5 would be identical to this one because of the way GitHub's UI presents the changes.

To review this PR is small chunks, I have to wait for 1/5 to be merged and then create 2/5 and so on. I would do that but is is just a lot of work for everyone :) So I suggest closing 1/5 and merging this one instead. What are your thoughts on this?

**Comment by sidcha** [comment]
I've added @adakus and @r2r0 as collaborators for this subsystem after getting their consent over email.

**Comment by str4t0m** [comment]
> @MaureenHelm I don't think splitting this PR is of much help;

I too prefer smaller PRs. If everything is crammed into a single PR errors are overlooked much more likely.
Also, sometimes collaborators/maintainers might have a few minutes available to squeeze in a PR review, but not enough to review a huge PR, leading to much quicker response times for smaller PRs.

In your current split of 5 PRs the fixes, API changes/additions, maintainer changes, and cleanups are still mixed together.
Why not create a PR with all/most of the cleanups and fixes first in a separate PR and API changes separately?
Also, I'd suggest to split the maintainer changes into a separate PR. There aren't any dependencies to other commits at all.
And in case the copyright changes block a merge, I'd suggest to leave them for now. Because they might even be merged after the merge window has closed. Although I am not completely sure about that.

**Comment by sidcha** [comment]
@str4t0m,

> I too prefer smaller PRs. If everything is crammed into a single PR errors are overlooked much more likely. Also, sometimes collaborators/maintainers might have a few minutes available to squeeze in a PR review, but not enough to review a huge PR, leading to much quicker response times for smaller PRs.

I agree with this sentiment overall, truly. As I explained in the dev-review meeting, the reasons for such a large PRs were:
- this is a series of patches developed over a year in another port of this module. I wasn't made aware of any active users so cleanups and enhancements did not reach here (what is already merged is standalone and complete).
- this subsystem did not have dedicated reviewers/collaborators so getting changes merged was cumbersome. Now we have fixed that. Moving forward I'll make sure to send out smaller PRs

Overall, the request was to treat this as a special case and bear with me :) 

> [..] Why not create a PR with all/most of the cleanups and fixes first in a separate PR and API changes separately?

I tried that, It increases the rebase effort significantly. I might as well create-wait-merge-repeat 5 separate PRs.

> [..] Also, I'd suggest to split the maintainer changes into a separate PR.

This was easy. Created #49303 with maintainers and other ad-hoc changes.

> [..] And in case the copyright changes block a merge, I'd suggest to leave them for now.

Figured so, the change is question is already dropped in current version.

I'll close this PR in favour of the smaller ones since it isn't going anywhere.

**Pull request closed without merging**
